src/citrine/builders/predictors.py

```diff
@@ -0,0 +1,125 @@
+    MeanPropertyPredictor
+)
+from citrine.informatics.descriptors import Descriptor, FormulationDescriptor, RealDescriptor
+
+
+def mean_feature_properties(
+        *,
+        project: Project,
+        featurizer: MolecularStructureFeaturizer,
+        formulation_descriptor: FormulationDescriptor,
+        p: int,
+        impute_properties: bool = True,
+        all_ingredients: bool = True,
+        labels: Optional[List[str]] = None
+) -> Tuple[List[MeanPropertyPredictor], List[Descriptor]]:
+    """[ALPHA] Combine a featurizer model with mean property models.
+
+    Given a featurizer, produce "mean property" models that calculate the mean of the
+    properties calculated by the featurizer. This is useful if you do not directly know
+    the numerical properties of ingredients in a formulation, but instead know, for example,
+    the molecular structure. This builder method determines the real-valued properties that the
+    featurizer calculates and builds mean property models that use them as input properties.
+
+    Parameters
+    ----------
+    project: Project
+        Project that contains the predictor
+    featurizer: MolecularStructureFeaturizer
+        A model that is being used to featurize formulation ingredients. Currently only
+        accepts a molecular structure featurizer, but other models can be used in principle.
+        TODO: add chemical formula featurizer once it is exposed
+    formulation_descriptor: FormulationDescriptor
+        Descriptor that represents the formulation being featurized.
+    p: int
+        Power of the generalized mean. Only integer powers are supported.
+    impute_properties: bool
+        Whether to impute missing ingredient properties by averaging over the entire dataset.
+        If ``False``, an error is thrown if any ingredient being featurized is missing a property.
+    all_ingredients: bool
+        Whether to create a MeanPropertyPredictor that calculates the mean over all ingredients.
+    labels: Optional[List[str]]
+        List of labels to create a MeanPropertyPredictor for.
+
+    Returns
+    -------
+    Tuple[List[MeanPropertyPredictor], List[Descriptor]]
+        List of mean property predictors that should be incorporated into the graph predictor,
+        and a list of all the output descriptors produced by these models. There will be one
+        model for each label specified, and one model for all ingredients if ``all_ingredients``
+        is set to ``True``. In the common case, the output descriptors will all be used
+        as inputs to one or more ML models.
+
+    """
+    if isinstance(featurizer, MolecularStructureFeaturizer):
+        input_descriptor = featurizer.descriptor
+    else:
+        raise TypeError(f"Featurizer of type {type(featurizer)} is not supported.")
+
+    if labels is None:
+        labels = []
+    if all_ingredients:
+        labels = [None] + labels
+    if not isinstance(labels, list) or len(labels) == 0:
```

Comment:
Based on the error message, I would have expected the following code:

```suggestion
    if not all_ingredients or len(labels) == 0:
```

Reply:
It's not an exclusive or. I'll change the message. This logic is doing two things -- it's making sure that `all_ingredients == true` or labels are specified, and it's also make sure that `labels` was specified as a list and not something else.